### ISSUE TITLE
Dask distributed does not run SVD on chunked data array

### ISSUE THREAD

**Issue by r-beer** [dask/dask#5368]
First of all, thank you very much for the helpful guidelines concerning bug reports. I have already learned a lot by just creating the mcve as explained in [Craft Minimal Bug Reports ](http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

I created this bug report as suggested by @mrocklin on this (less clean) [related stackoverflow question](https://stackoverflow.com/questions/57771434/dask-distributed-does-not-run-svd-if-some-of-the-chunks-contain-only-nan-values).

When trying to run code equivalent to the following mcve:

```python
# imports
import numpy as np
import pandas as pd
import xarray as xr
from dask.distributed import Client
from dask_jobqueue import SLURMCluster
from eofs.xarray import Eof as xEof

# create data
n_eofs = 100
raster = np.zeros((200, 200))
raster[75:125, 75:125] = 1
np_data = np.random.rand(60, 200, 200)
x = [i for i in np.arange(200)]
y = [i for i in np.arange(200)]
times = pd.date_range('2000-01-01', periods=60)
data = xr.DataArray(np_data, coords=[times, x, y], dims=['time', 'x', 'y'])
data_masked = data.where(raster)

# chunk data
chunk_dict = {'x': 20, 'y': 20}
data = data.chunk(chunks=chunk_dict)
data_masked = data_masked.chunk(chunks=chunk_dict)

# run dask client
client = Client()

# chunk data
solver = xEof(data)
eofs = solver.eofsAsCovariance(neofs=n_eofs)
pcs = solver.pcs(npcs=n_eofs, pcscaling=1)
```
I get the following error:

```python-traceback
ValueError: error encountered in SVD, check that missing values are in the same places at each time and that all the values are not missing
```

<details>

```python-traceback

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~/.conda/envs/spagat_py36/lib/python3.6/site-packages/eofs/standard.py in __init__(self, dataset, weights, center, ddof)
    164                 # Use the parallel Dask algorithm
--> 165                 dsvd = dask.array.linalg.svd(dataNoMissing)
    166                 A, Lh, E = (x.compute() for x in dsvd)

~/.conda/envs/spagat_py36/lib/python3.6/site-packages/dask/array/linalg.py in svd(a)
    803     """
--> 804     return tsqr(a, compute_svd=True)
    805 

~/.conda/envs/spagat_py36/lib/python3.6/site-packages/dask/array/linalg.py in tsqr(data, compute_svd, _max_vchunk_size)
    116             "Current shape: {},\nCurrent chunksize: {}".format(
--> 117                 data.shape, data.chunksize
    118             )

ValueError: Input must have the following properties:
  1. Have two dimensions
  2. Have only one column of blocks

Note: This function (tsqr) supports QR decomposition in the case of
tall-and-skinny matrices (single column chunk/block; see qr)Current shape: (60, nan),
Current chunksize: (60, nan)

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-15-7cd684c0ba85> in <module>
----> 1 solver = xEof(data)
      2 eofs = solver.eofsAsCovariance(neofs=n_eofs)
      3 pcs = solver.pcs(npcs=n_eofs, pcscaling=1)

~/.conda/envs/spagat_py36/lib/python3.6/site-packages/eofs/xarray.py in __init__(self, array, weights, center, ddof)
    131                                     weights=wtarray,
    132                                     center=center,
--> 133                                     ddof=ddof)
    134         # Name of the input DataArray.
    135         self._name = array.name

~/.conda/envs/spagat_py36/lib/python3.6/site-packages/eofs/standard.py in __init__(self, dataset, weights, center, ddof)
    175 
    176         except (np.linalg.LinAlgError, ValueError):
--> 177             raise ValueError('error encountered in SVD, check that missing '
    178                              'values are in the same places at each time and '
    179                              'that all the values are not missing')

ValueError: error encountered in SVD, check that missing values are in the same places at each time and that all the values are not missing
```
</details>

As the [xeof analysis](https://ajdawson.github.io/eofs/latest/api/eofs.xarray.html#module-eofs.xarray) works fine when using the exact same data without chunking, this might be labeled as bug of `dask.distributed`. For further information on that please find attached [test_chunked_xeof.txt](https://github.com/dask/dask/files/3585430/test_chunked_xeof.txt), a few small pytest-compatible testing functions that might help you for debugging. 

PS: I am still new to contributing open source code, so I am happy for any critical feedback concerning the way I contribute.

**Comment by TomAugspurger** [comment]
> PS: I am still new to contributing open source code, so I am happy for any critical feedback concerning the way I contribute.

I'll start here :)

1. It looks like dask-jobqueu / SLURMCluster isn't needed
2. It appears that creating a Client isn't needed. Unless the bug is in the distributed scheduler, you'll find that dask issues can often be reproduced without it. You'll learn this in time :)
3. Dig a bit into what `eof` is doing (this is harder).

I happen to know that dask's current `svd` implementation only works for arrays chunked along the first dimension. it's only appropriate for tall and skinny matrices.

```python
In [1]: import dask.array as da

In [2]: X = da.random.random((10, 10), chunks=(5, 5))

In [3]: da.linalg.svd(X)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-375edcc7942c> in <module>
----> 1 da.linalg.svd(X)

~/sandbox/dask/dask/array/linalg.py in svd(a)
    802     dask.array.linalg.tsqr: Implementation for tall-and-skinny arrays
    803     """
--> 804     return tsqr(a, compute_svd=True)
    805
    806

~/sandbox/dask/dask/array/linalg.py in tsqr(data, compute_svd, _max_vchunk_size)
    115             "tall-and-skinny matrices (single column chunk/block; see qr)"
    116             "Current shape: {},\nCurrent chunksize: {}".format(
--> 117                 data.shape, data.chunksize
    118             )
    119         )

ValueError: Input must have the following properties:
  1. Have two dimensions
  2. Have only one column of blocks

Note: This function (tsqr) supports QR decomposition in the case of
tall-and-skinny matrices (single column chunk/block; see qr)Current shape: (10, 10),
Current chunksize: (5, 5)
```

This is present in your traceback, though it's a bit burried. You might open an issue on the eofs tracker suggesting they do

```python
except (np.linalg.LinAlgError, ValueError) as e:
    raise ... from e
```

but that's minor.

So the solution, if feasible for your dataset, is to only chunk along the first dimension. Something like 


```
chunk_dict = {"time": n}
```
for some appropriate `n`.

**Comment by r-beer** [comment]
Dear @TomAugspurger, 
thank you very much for your fast reply!

It took me a while to verify whether this really solves the chunking problem when using `dask.distributed`. It does, thank you very much! 👍 

> 1. It looks like dask-jobqueu / SLURMCluster isn't needed

Exactly, I didn't properly clean up the mcve.

> 2. It appears that creating a Client isn't needed. Unless the bug is in the distributed scheduler, you'll find that dask issues can often be reproduced without it. You'll learn this in time :)

Thank you for the info, that indeed facilitates debugging. Especially when braking down the mcve further, as you show.

> 3. Dig a bit into what eof is doing (this is harder).

Thank you for your explanations!

**tl;dr:**
Finally, I would like to give you the chance of repeating your answer on stack overflow, if you want to. If you [paste your answer here](https://stackoverflow.com/questions/57771434/dask-distributed-does-not-run-svd-if-some-of-the-chunks-contain-only-nan-values), then I will kindly accept it. 

Otherwise, I will close the SO question in a few days, forwarding to this discussion.
